{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "import pandas as pd\n",
    "import gumpy\n",
    "import parse\n",
    "\n",
    "pd.options.display.max_columns=999"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 8336/8336 [00:00<00:00, 209541.75it/s]\n",
      "100%|██████████| 3909/3909 [00:48<00:00, 80.66it/s]\n",
      "100%|██████████| 100/100 [00:01<00:00, 72.57it/s]\n"
     ]
    }
   ],
   "source": [
    "reference = gumpy.Genome(\"../gumpy/config/NC_000962.3.gbk.gz\", show_progress_bar=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_who_catalog(filename):\n",
    "    '''Parses the WHO TB catalog\n",
    "\n",
    "    Args:\n",
    "        filename (str): Path to the WHO catalog\n",
    "    Returns:\n",
    "        pd.dataframe: Dataframe containing the mutations\n",
    "    '''\n",
    "    df = pd.read_excel(filename, sheet_name=\"Genome_indices\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load the catalogue\n",
    "data = parse_who_catalog(\"WHO-UCN-GTB-PCI-2021.7-eng.xlsx\")\n",
    "\n",
    "#The catalogue is grouped by gene, so we can store gene masks until they are no longer required\n",
    "masks = {}\n",
    "\n",
    "#Setup details for drugs\n",
    "drug_columns = [\n",
    "    'RIF_Conf_Grade',\n",
    "    'INH_Conf_Grade',\n",
    "    'EMB_Conf_Grade',\n",
    "    'PZA_Conf_Grade',\n",
    "    'LEV_Conf_Grade',\n",
    "    'MXF_Conf_Grade',\n",
    "    'BDQ_Conf_Grade',\n",
    "    'LZD_Conf_Grade',\n",
    "    'CFZ_Conf_Grade',\n",
    "    'DLM_Conf_Grade',\n",
    "    'AMI_Conf_Grade',\n",
    "    'STM_Conf_Grade',\n",
    "    'ETH_Conf_Grade',\n",
    "    'KAN_Conf_Grade',\n",
    "    'CAP_Conf_Grade']\n",
    "drugs = {\n",
    "    drug.split(\"_\")[0]: {\n",
    "        \"R\": set(),\n",
    "        \"U\": set(),\n",
    "        \"S\": set(),\n",
    "        # \"F\": set()\n",
    "        } for drug in drug_columns}\n",
    "genes = set()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_locus</th>\n",
       "      <th>variant</th>\n",
       "      <th>codon_number</th>\n",
       "      <th>genome_index</th>\n",
       "      <th>ref_nt</th>\n",
       "      <th>alt_nt</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>Total_R</th>\n",
       "      <th>Total_S</th>\n",
       "      <th>RIF_R</th>\n",
       "      <th>RIF_S</th>\n",
       "      <th>RIF_Conf_Grade</th>\n",
       "      <th>INH_R</th>\n",
       "      <th>INH_S</th>\n",
       "      <th>INH_Conf_Grade</th>\n",
       "      <th>EMB_R</th>\n",
       "      <th>EMB_S</th>\n",
       "      <th>EMB_Conf_Grade</th>\n",
       "      <th>PZA_R</th>\n",
       "      <th>PZA_S</th>\n",
       "      <th>PZA_Conf_Grade</th>\n",
       "      <th>LEV_R</th>\n",
       "      <th>LEV_S</th>\n",
       "      <th>LEV_Conf_Grade</th>\n",
       "      <th>MXF_R</th>\n",
       "      <th>MXF_S</th>\n",
       "      <th>MXF_Conf_Grade</th>\n",
       "      <th>BDQ_R</th>\n",
       "      <th>BDQ_S</th>\n",
       "      <th>BDQ_Conf_Grade</th>\n",
       "      <th>LZD_R</th>\n",
       "      <th>LZD_S</th>\n",
       "      <th>LZD_Conf_Grade</th>\n",
       "      <th>CFZ_R</th>\n",
       "      <th>CFZ_S</th>\n",
       "      <th>CFZ_Conf_Grade</th>\n",
       "      <th>DLM_R</th>\n",
       "      <th>DLM_S</th>\n",
       "      <th>DLM_Conf_Grade</th>\n",
       "      <th>AMI_R</th>\n",
       "      <th>AMI_S</th>\n",
       "      <th>AMI_Conf_Grade</th>\n",
       "      <th>STM_R</th>\n",
       "      <th>STM_S</th>\n",
       "      <th>STM_Conf_Grade</th>\n",
       "      <th>ETH_R</th>\n",
       "      <th>ETH_S</th>\n",
       "      <th>ETH_Conf_Grade</th>\n",
       "      <th>KAN_R</th>\n",
       "      <th>KAN_S</th>\n",
       "      <th>KAN_Conf_Grade</th>\n",
       "      <th>CAP_R</th>\n",
       "      <th>CAP_S</th>\n",
       "      <th>CAP_Conf_Grade</th>\n",
       "      <th>final_annotation.Reference</th>\n",
       "      <th>final_annotation.Position</th>\n",
       "      <th>final_annotation.ReferenceNucleotide</th>\n",
       "      <th>final_annotation.AlternativeNucleotide</th>\n",
       "      <th>final_annotation.Gene</th>\n",
       "      <th>final_annotation.Effect</th>\n",
       "      <th>final_annotation.LocusTag</th>\n",
       "      <th>final_annotation.TentativeHGVSNucleotidicAnnotation</th>\n",
       "      <th>final_annotation.ProteinId</th>\n",
       "      <th>final_annotation.TentativeHGVSProteicAnnotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB_c-170t</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5070</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>5070</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>c.-170C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB_-165_del_1_cg_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5075</td>\n",
       "      <td>cg</td>\n",
       "      <td>c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>5075</td>\n",
       "      <td>cg</td>\n",
       "      <td>c</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>c.-164delG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB_-165_ins_2_cg_cgtc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5075</td>\n",
       "      <td>cg</td>\n",
       "      <td>cgtc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>5075</td>\n",
       "      <td>cg</td>\n",
       "      <td>cgtc</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>c.-164_-163insTC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB_c-165t</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5075</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>793</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>423.0</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>4.0</td>\n",
       "      <td>370.0</td>\n",
       "      <td>5) Not assoc w R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>5075</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>c.-165C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB_-165_ins_1_cg_cgt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5075</td>\n",
       "      <td>cg</td>\n",
       "      <td>cgt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>3) Uncertain significance</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>5075</td>\n",
       "      <td>cg</td>\n",
       "      <td>cgt</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>c.-164_-163insT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name gene_locus                  variant codon_number genome_index  \\\n",
       "0      gyrB     Rv0005              gyrB_c-170t          NaN         5070   \n",
       "1      gyrB     Rv0005     gyrB_-165_del_1_cg_c          NaN         5075   \n",
       "2      gyrB     Rv0005  gyrB_-165_ins_2_cg_cgtc          NaN         5075   \n",
       "3      gyrB     Rv0005              gyrB_c-165t          NaN         5075   \n",
       "4      gyrB     Rv0005   gyrB_-165_ins_1_cg_cgt          NaN         5075   \n",
       "\n",
       "  ref_nt alt_nt ref_aa alt_aa  Total_R  Total_S  RIF_R  RIF_S RIF_Conf_Grade  \\\n",
       "0      c      t    NaN    NaN        0        2    NaN    NaN            NaN   \n",
       "1     cg      c    NaN    NaN        0        4    NaN    NaN            NaN   \n",
       "2     cg   cgtc    NaN    NaN        0        4    NaN    NaN            NaN   \n",
       "3      c      t    NaN    NaN       13      793    NaN    NaN            NaN   \n",
       "4     cg    cgt    NaN    NaN        0       72    NaN    NaN            NaN   \n",
       "\n",
       "   INH_R  INH_S INH_Conf_Grade  EMB_R  EMB_S EMB_Conf_Grade  PZA_R  PZA_S  \\\n",
       "0    NaN    NaN            NaN    NaN    NaN            NaN    NaN    NaN   \n",
       "1    NaN    NaN            NaN    NaN    NaN            NaN    NaN    NaN   \n",
       "2    NaN    NaN            NaN    NaN    NaN            NaN    NaN    NaN   \n",
       "3    NaN    NaN            NaN    NaN    NaN            NaN    NaN    NaN   \n",
       "4    NaN    NaN            NaN    NaN    NaN            NaN    NaN    NaN   \n",
       "\n",
       "  PZA_Conf_Grade  LEV_R  LEV_S             LEV_Conf_Grade  MXF_R  MXF_S  \\\n",
       "0            NaN    0.0    1.0  3) Uncertain significance    0.0    1.0   \n",
       "1            NaN    0.0    3.0  3) Uncertain significance    0.0    1.0   \n",
       "2            NaN    0.0    2.0  3) Uncertain significance    0.0    2.0   \n",
       "3            NaN    9.0  423.0           5) Not assoc w R    4.0  370.0   \n",
       "4            NaN    0.0   38.0  3) Uncertain significance    0.0   34.0   \n",
       "\n",
       "              MXF_Conf_Grade  BDQ_R  BDQ_S BDQ_Conf_Grade  LZD_R  LZD_S  \\\n",
       "0  3) Uncertain significance    NaN    NaN            NaN    NaN    NaN   \n",
       "1  3) Uncertain significance    NaN    NaN            NaN    NaN    NaN   \n",
       "2  3) Uncertain significance    NaN    NaN            NaN    NaN    NaN   \n",
       "3           5) Not assoc w R    NaN    NaN            NaN    NaN    NaN   \n",
       "4  3) Uncertain significance    NaN    NaN            NaN    NaN    NaN   \n",
       "\n",
       "  LZD_Conf_Grade  CFZ_R  CFZ_S CFZ_Conf_Grade  DLM_R  DLM_S DLM_Conf_Grade  \\\n",
       "0            NaN    NaN    NaN            NaN    NaN    NaN            NaN   \n",
       "1            NaN    NaN    NaN            NaN    NaN    NaN            NaN   \n",
       "2            NaN    NaN    NaN            NaN    NaN    NaN            NaN   \n",
       "3            NaN    NaN    NaN            NaN    NaN    NaN            NaN   \n",
       "4            NaN    NaN    NaN            NaN    NaN    NaN            NaN   \n",
       "\n",
       "   AMI_R  AMI_S AMI_Conf_Grade  STM_R  STM_S STM_Conf_Grade  ETH_R  ETH_S  \\\n",
       "0    NaN    NaN            NaN    NaN    NaN            NaN    NaN    NaN   \n",
       "1    NaN    NaN            NaN    NaN    NaN            NaN    NaN    NaN   \n",
       "2    NaN    NaN            NaN    NaN    NaN            NaN    NaN    NaN   \n",
       "3    NaN    NaN            NaN    NaN    NaN            NaN    NaN    NaN   \n",
       "4    NaN    NaN            NaN    NaN    NaN            NaN    NaN    NaN   \n",
       "\n",
       "  ETH_Conf_Grade  KAN_R  KAN_S KAN_Conf_Grade  CAP_R  CAP_S CAP_Conf_Grade  \\\n",
       "0            NaN    NaN    NaN            NaN    NaN    NaN            NaN   \n",
       "1            NaN    NaN    NaN            NaN    NaN    NaN            NaN   \n",
       "2            NaN    NaN    NaN            NaN    NaN    NaN            NaN   \n",
       "3            NaN    NaN    NaN            NaN    NaN    NaN            NaN   \n",
       "4            NaN    NaN    NaN            NaN    NaN    NaN            NaN   \n",
       "\n",
       "  final_annotation.Reference  final_annotation.Position  \\\n",
       "0                NC_000962.3                       5070   \n",
       "1                NC_000962.3                       5075   \n",
       "2                NC_000962.3                       5075   \n",
       "3                NC_000962.3                       5075   \n",
       "4                NC_000962.3                       5075   \n",
       "\n",
       "  final_annotation.ReferenceNucleotide final_annotation.AlternativeNucleotide  \\\n",
       "0                                    c                                      t   \n",
       "1                                   cg                                      c   \n",
       "2                                   cg                                   cgtc   \n",
       "3                                    c                                      t   \n",
       "4                                   cg                                    cgt   \n",
       "\n",
       "  final_annotation.Gene final_annotation.Effect final_annotation.LocusTag  \\\n",
       "0                  gyrB   upstream_gene_variant                    Rv0005   \n",
       "1                  gyrB   upstream_gene_variant                    Rv0005   \n",
       "2                  gyrB   upstream_gene_variant                    Rv0005   \n",
       "3                  gyrB   upstream_gene_variant                    Rv0005   \n",
       "4                  gyrB   upstream_gene_variant                    Rv0005   \n",
       "\n",
       "  final_annotation.TentativeHGVSNucleotidicAnnotation  \\\n",
       "0                                          c.-170C>T    \n",
       "1                                         c.-164delG    \n",
       "2                                   c.-164_-163insTC    \n",
       "3                                          c.-165C>T    \n",
       "4                                    c.-164_-163insT    \n",
       "\n",
       "  final_annotation.ProteinId final_annotation.TentativeHGVSProteicAnnotation  \n",
       "0                        NaN                                             NaN  \n",
       "1                        NaN                                             NaN  \n",
       "2                        NaN                                             NaN  \n",
       "3                        NaN                                             NaN  \n",
       "4                        NaN                                             NaN  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5070 c t\n",
      "5075 cg c\n",
      "5075 cg cgtc\n",
      "5075 c t\n",
      "5075 cg cgt\n"
     ]
    }
   ],
   "source": [
    "for (index, row) in data[:5].iterrows():\n",
    "    garc = []\n",
    "    #Pull out gene name, pos, ref and alt\n",
    "    gene = row[\"gene_name\"]\n",
    "    genes.add(gene)\n",
    "    if masks.get(gene) is None:\n",
    "        #Cache the masks\n",
    "        masks = {gene: parse.get_masks(reference, gene)}\n",
    "    pos = str(row[\"final_annotation.Position\"])#Cast to a str for str.split(',')\n",
    "    ref = row[\"final_annotation.ReferenceNucleotide\"]\n",
    "    alt = row[\"final_annotation.AlternativeNucleotide\"]\n",
    "\n",
    "    #Check for multiple positions defined within pos\n",
    "    if len(pos.split(\",\")) > 1:\n",
    "        #There is more than 1 mutation detailed in this row, so skip it\n",
    "        print(\"Mulitple muations per row: \", gene, pos, ref, alt, sep=\"\\t\")\n",
    "        continue\n",
    "    else:            \n",
    "        garc += parse.to_garc(reference, gene, int(pos), ref, alt, masks)\n",
    "        if len(garc) > 1:\n",
    "            #There is more than 1 mutation generated from this row, so skip it\n",
    "            print(\"Multiple mutations per row: \", gene, pos, ref, alt, garc, sep=\"\\t\")\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gyrB': (array([[False, False, False, ..., False, False, False],\n",
       "         [False, False, False, ..., False, False, False]]),\n",
       "  array([False, False, False, ..., False, False, False]))}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "masks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "7a792fcb311f9eb9f3c1b942a8c87ada8484712b89b670347c16a1088e0a1f69"
  },
  "kernelspec": {
   "display_name": "Python 3.10.4 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
